src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs

```diff
@@ -30,14 +30,58 @@ public sealed class CustomResourceAnnotation(Func<CustomResourceSnapshot> initia
     /// Updates the snapshot of the <see cref="CustomResourceSnapshot"/> for a resource.
     /// </summary>
     /// <param name="state">The new <see cref="CustomResourceSnapshot"/>.</param>
-    public async Task UpdateStateAsync(CustomResourceSnapshot state)
+    public Task UpdateStateAsync(CustomResourceSnapshot state)
+    {
+        if (_streamClosedCts.IsCancellationRequested)
+        {
+            return Task.CompletedTask;
+        }
+
+        OnSnapshotUpdated?.Invoke(state);
+
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Signal that no more updates are expected for this resource.
+    /// </summary>
+    public void Complete()
+    {
+        _streamClosedCts.Cancel();
+    }
+
+    private sealed class ResourceUpdatesAsyncEnumerable(ResourceUpdatesAnnotation customResourceAnnotation) : IAsyncEnumerable<CustomResourceSnapshot>
     {
-        await _channel.Writer.WriteAsync(state).ConfigureAwait(false);
+        public async IAsyncEnumerator<CustomResourceSnapshot> GetAsyncEnumerator(CancellationToken cancellationToken = default)
+        {
+            var channel = Channel.CreateUnbounded<CustomResourceSnapshot>();
+
+            void WriteToChannel(CustomResourceSnapshot state)
+                => channel.Writer.TryWrite(state);
+
+            using var _ = customResourceAnnotation._streamClosedCts.Token.Register(() => channel.Writer.TryComplete());
+
+            customResourceAnnotation.OnSnapshotUpdated = WriteToChannel;
+
+            try
+            {
+                await foreach (var item in channel.Reader.ReadAllAsync(cancellationToken))
+                {
+                    yield return item;
+                }
+            }
+            finally
+            {
+                customResourceAnnotation.OnSnapshotUpdated -= WriteToChannel;
+
+                channel.Writer.TryComplete();
+            }
+        }
     }
 }
 
 /// <summary>
-/// The context for a all of the properties and URLs that should show up in the dashboard for a resource.
+/// An immutable snapshot of the state of a resource.
 /// </summary>
 public sealed record CustomResourceSnapshot
 {
```

Comment:
A few members down, `EnviomentVariables` is mis-spelled, it should be `EnvionmentVariables`